### ISSUE TITLE
Failed to compile on Mac OS Mojave 10.14.5 (18F132)

### ISSUE THREAD

**Issue by newmacuser611** [nschloe/accupy#11]
I tried installing accupy and compilation failed:
```  
~/partitioning: python3 -m pip install accupy
Collecting accupy
...
    Complete output from command /usr/local/bin/python3 -u -c "import setuptools, tokenize;__file__='/private/var/folders/_j/_5vgryt961bg8tw_2jvgmc6w0000gp/T/pip-install-t2i_vxvy/pyfma/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace([87/1791]
\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /private/var/folders/_j/_5vgryt961bg8tw_2jvgmc6w0000gp/T/pip-record-38rgrthc/install-record.txt --single-version-externally-managed --compile:
    running install
    running build
    running build_py
    creating build
    creating build/lib.macosx-10.9-x86_64-3.7
    creating build/lib.macosx-10.9-x86_64-3.7/pyfma
    copying pyfma/__init__.py -> build/lib.macosx-10.9-x86_64-3.7/pyfma
    copying pyfma/__about__.py -> build/lib.macosx-10.9-x86_64-3.7/pyfma
    running build_ext
    building '_pyfma' extension
    creating build/temp.macosx-10.9-x86_64-3.7
    creating build/temp.macosx-10.9-x86_64-3.7/src
    gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -arch x86_64 -g -I/Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m -I/Users/smukherjee/Library/Python/3.7/include/python3.7m -I/Library/Fr
ameworks/Python.framework/Versions/3.7/include/python3.7m -c src/pybind11.cpp -o build/temp.macosx-10.9-x86_64-3.7/src/pybind11.o
    In file included from src/pybind11.cpp:3:
    In file included from /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/pybind11.h:44:
    In file included from /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/attr.h:13:
    In file included from /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/cast.h:13:
    In file included from /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/pytypes.h:12:
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:298:17: warning: alias declarations are a C++11 extension [-Wc++11-extensions]
    using ssize_t = Py_ssize_t;
                    ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:299:17: warning: alias declarations are a C++11 extension [-Wc++11-extensions]
    using size_t  = std::size_t;
                    ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:302:6: warning: scoped enumerations are a C++11 extension [-Wc++11-extensions]
    enum class return_value_policy : uint8_t {
         ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:302:12: error: ISO C++ forbids forward references to 'enum' types
    enum class return_value_policy : uint8_t {
               ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:302:32: error: expected unqualified-id
    enum class return_value_policy : uint8_t {
                                   ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:355:15: error: unknown type name 'constexpr'
    inline static constexpr int log2(size_t n, int k = 0) { return (n <= 1) ? k : log2(n >> 1, k + 1); }
                  ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:355:25: error: expected unqualified-id
    inline static constexpr int log2(size_t n, int k = 0) { return (n <= 1) ? k : log2(n >> 1, k + 1); }
                            ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:358:15: error: unknown type name 'constexpr'
    inline static constexpr size_t size_in_ptrs(size_t s) { return 1 + ((s - 1) >> log2(sizeof(void *))); }
                  ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:358:1: warning: inline variables are a C++17 extension [-Wc++17-extensions]
    inline static constexpr size_t size_in_ptrs(size_t s) { return 1 + ((s - 1) >> log2(sizeof(void *))); }
    ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:358:31: error: expected ';' after top level declarator
    inline static constexpr size_t size_in_ptrs(size_t s) { return 1 + ((s - 1) >> log2(sizeof(void *))); }
                                  ^
                                  ;
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:386:39: error: use of undeclared identifier 'instance_simple_holder_in_ptrs'
            void *simple_value_holder[1 + instance_simple_holder_in_ptrs()];
                                          ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:433:49: error: unknown type name 'type_info'; did you mean 'std::type_info'?
        value_and_holder get_value_and_holder(const type_info *find_type = nullptr, bool throw_if_missing = true);
                                                    ^~~~~~~~~
                                                    std::type_info
    /Library/Developer/CommandLineTools/usr/include/c++/v1/typeinfo:87:29: note: 'std::type_info' declared here                                                                                                                                                      [28/1791]
    class _LIBCPP_EXCEPTION_ABI type_info
                                ^
    In file included from src/pybind11.cpp:3:
    In file included from /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/pybind11.h:44:
    In file included from /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/attr.h:13:
    In file included from /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/cast.h:13:
    In file included from /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/pytypes.h:12:
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:436:12: error: unknown type name 'constexpr'
        static constexpr uint8_t status_holder_constructed  = 1;
               ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:436:29: error: expected ';' at end of declaration list
        static constexpr uint8_t status_holder_constructed  = 1;
                                ^
                                ;
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:437:12: error: unknown type name 'constexpr'
        static constexpr uint8_t status_instance_registered = 2;
               ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:437:29: error: expected ';' at end of declaration list
        static constexpr uint8_t status_instance_registered = 2;
                                ^
                                ;
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:449:58: warning: alias declarations are a C++11 extension [-Wc++11-extensions]
    template <bool B, typename T = void> using enable_if_t = typename std::enable_if<B, T>::type;
                                                             ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:450:65: warning: alias declarations are a C++11 extension [-Wc++11-extensions]
    template <bool B, typename T, typename F> using conditional_t = typename std::conditional<B, T, F>::type;
                                                                    ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:451:43: warning: alias declarations are a C++11 extension [-Wc++11-extensions]
    template <typename T> using remove_cv_t = typename std::remove_cv<T>::type;
                                              ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:452:50: warning: alias declarations are a C++11 extension [-Wc++11-extensions]
    template <typename T> using remove_reference_t = typename std::remove_reference<T>::type;
                                                     ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:460:10: error: unknown type name 'size_t'
    template<size_t ...> struct index_sequence  { };
             ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:460:17: warning: variadic templates are a C++11 extension [-Wc++11-extensions]
    template<size_t ...> struct index_sequence  { };
                    ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:461:10: error: unknown type name 'size_t'
    template<size_t N, size_t ...S> struct make_index_sequence_impl : make_index_sequence_impl <N - 1, N - 1, S...> { };
             ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:461:20: error: unknown type name 'size_t'
    template<size_t N, size_t ...S> struct make_index_sequence_impl : make_index_sequence_impl <N - 1, N - 1, S...> { };
                       ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:461:27: warning: variadic templates are a C++11 extension [-Wc++11-extensions]
    template<size_t N, size_t ...S> struct make_index_sequence_impl : make_index_sequence_impl <N - 1, N - 1, S...> { };
                              ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:462:10: error: unknown type name 'size_t'
    template<size_t ...S> struct make_index_sequence_impl <0, S...> { typedef index_sequence<S...> type; };
             ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:462:17: warning: variadic templates are a C++11 extension [-Wc++11-extensions]
    template<size_t ...S> struct make_index_sequence_impl <0, S...> { typedef index_sequence<S...> type; };
                    ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:463:10: error: unknown type name 'size_t'
    template<size_t N> using make_index_sequence = typename make_index_sequence_impl<N>::type;
             ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:463:48: warning: alias declarations are a C++11 extension [-Wc++11-extensions]
    template<size_t N> using make_index_sequence = typename make_index_sequence_impl<N>::type;
                                                   ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:467:32: error: expected template parameter
    template <typename ISeq, size_t, bool...> struct select_indices_impl { using type = ISeq; };
                                   ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:467:38: warning: variadic templates are a C++11 extension [-Wc++11-extensions]
    template <typename ISeq, size_t, bool...> struct select_indices_impl { using type = ISeq; };
                                         ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:467:85: warning: alias declarations are a C++11 extension [-Wc++11-extensions]
    template <typename ISeq, size_t, bool...> struct select_indices_impl { using type = ISeq; };
                                                                                        ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:468:11: error: unknown type name 'size_t'
    template <size_t... IPrev, size_t I, bool B, bool... Bs> struct select_indices_impl<index_sequence<IPrev...>, I, B, Bs...>
              ^
    /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/pybind11/detail/common.h:468:17: warning: variadic templates are a C++11 extension [-Wc++11-extensions]
    template <size_t... IPrev, size_t I, bool B, bool... Bs> struct select_indices_impl<index_sequence<IPrev...>, I, B, Bs...>
                    ^
    fatal error: too many errors emitted, stopping now [-ferror-limit=]
    15 warnings and 20 errors generated.
    error: command 'gcc' failed with exit status 1
```

It looks like one of the problems is that my compiler is not expecting C++11. I'm going to try to see if if I can pass a flag to gcc to get it to accept C++ 11.

**Comment by nschloe** [comment]
Yeah, this definitely looks like a compiler error. Perhaps upgrading to Catalina will already do the job.

**Comment by goracle** [comment]
I can confirm this fails to build on the KNL at Thomas Jefferson Lab.  As @suryamukherjee10 guessed, it needed -std=c++11.  I used only local installs via pip, so I also had to manually install pybind11 and add the headers to my CFLAGS.  It also complained about missing BLAS headers from Eigen/Core, which I had to add to CFLAGS.  I was installing via python3, so this may have been the issue.  Cheers.

**Comment by nschloe** [comment]
Thanks for the comment. You haven't mentioned what system you're running, but if your eigen/pybind installation doesn't put the include paths in in the standard search paths, then you'll have to do that yourself via CFLAGS; accupy can't do anything about it.

About the `c++11` requirement, I guess I _could_ try and figure out if such a flag exists with the current compiler. I'm pretty sure I don't want to invest time in that though. The thing is that the standard came out 8 years ago, GCC has c++11 as default since 3.5 years, and it's in all major Linux distros for a while now. If you're using a compiler older than that (or one that's behind schedule in other ways), then you'll have expect to do some extra work (like you did).

**Comment by goracle** [comment]
Well, I'm unsure if that's true.  I used pip to install pybind11 with the --user flag, so it seems to me accupy could look there in principle for the headers.  Eigen, I can see that that's true.  Major caveats to these comments, however, are in my next paragraph.

I did mention that this was a KNL (Intel knight's landing) running at a (US) national lab (with some info available online for those curious), so I did sort of broadly specify the system (but you might not be overly familiar with supercomputing environments).  The setup is, of course, a custom Linux system with deliberately (slightly) out of date (and likely custom built) compilers, so it's usually anyone's guess what aspects I would need to specify to help us debug the install there.  I agree that you shouldn't waste time supporting this kind of niche build, but this is the reason I wasn't more specific.  

I also want to plug some very general vector functionality I added to accupy which made it more useful for me. In: https://github.com/goracle/lattice-fitter/blob/master/latfit/utilities/exactmean.py.  I hope you might be interested in stealing these functions to add to accupy.  Feel free.  Thanks.
